res/controllers/Pioneer-DDJ-SB3-scripts.js

```diff
@@ -0,0 +1,1791 @@
+        }
+        if (PioneerDDJSB3.blinkAutodjState === 1) {
+            PioneerDDJSB3.valueVuMeter["[Channel1]_enabled"] = 1;
+            PioneerDDJSB3.valueVuMeter["[Channel3]_enabled"] = 1;
+            PioneerDDJSB3.valueVuMeter["[Channel2]_enabled"] = 0;
+            PioneerDDJSB3.valueVuMeter["[Channel4]_enabled"] = 0;
+        }
+        if (PioneerDDJSB3.blinkAutodjState === 2) {
+            PioneerDDJSB3.valueVuMeter["[Channel1]_enabled"] = 1;
+            PioneerDDJSB3.valueVuMeter["[Channel3]_enabled"] = 1;
+            PioneerDDJSB3.valueVuMeter["[Channel2]_enabled"] = 1;
+            PioneerDDJSB3.valueVuMeter["[Channel4]_enabled"] = 1;
+        }
+        if (PioneerDDJSB3.blinkAutodjState === 3) {
+            PioneerDDJSB3.valueVuMeter["[Channel1]_enabled"] = 0;
+            PioneerDDJSB3.valueVuMeter["[Channel3]_enabled"] = 0;
+            PioneerDDJSB3.valueVuMeter["[Channel2]_enabled"] = 1;
+            PioneerDDJSB3.valueVuMeter["[Channel4]_enabled"] = 1;
+        }
+    } else {
+        PioneerDDJSB3.valueVuMeter["[Channel1]_enabled"] = 1;
+        PioneerDDJSB3.valueVuMeter["[Channel3]_enabled"] = 1;
+        PioneerDDJSB3.valueVuMeter["[Channel2]_enabled"] = 1;
+        PioneerDDJSB3.valueVuMeter["[Channel4]_enabled"] = 1;
+    }
+};
+
+
+///////////////////////////////////////////////////////////////
+//                        AutoDJ                             //
+///////////////////////////////////////////////////////////////
+
+PioneerDDJSB3.autodjSkipNext = function(channel, control, value) {
+    if (value === 0) {
+        return;
+    }
+    if (engine.getValue("[AutoDJ]", "enabled")) {
+        engine.setValue("[AutoDJ]", "skip_next", true);
+    }
+};
+
+PioneerDDJSB3.autodjToggle = function(channel, control, value) {
+    if (value === 0) {
+        return;
+    }
+    if (engine.getValue("[AutoDJ]", "enabled")) {
+        engine.setValue("[AutoDJ]", "enabled", false);
+    } else {
+        engine.setValue("[AutoDJ]", "enabled", true);
+    }
+};
+
+
+///////////////////////////////////////////////////////////////
+//                      CONTROL BINDING                      //
+///////////////////////////////////////////////////////////////
+
+PioneerDDJSB3.bindSamplerControlConnections = function(samplerGroup) {
+    engine.makeConnection(samplerGroup, "duration", PioneerDDJSB3.samplerLedsDuration);
+    engine.makeConnection(samplerGroup, "play", PioneerDDJSB3.samplerLedsPlay);
+};
+
+PioneerDDJSB3.bindDeckControlConnections = function(channelGroup, isUnbinding) {
+    var i,
+        index,
+        controlsToFunctions = {
+            "pfl": "PioneerDDJSB3.headphoneCueLed",
+            "keylock": "PioneerDDJSB3.keyLockLed",
+            "loop_enabled": "PioneerDDJSB3.autoLoopLed",
+        };
+
+    controlsToFunctions.slipEnabled = "PioneerDDJSB3.slipLed";
+
+    for (i = 1; i <= 8; i++) {
+        controlsToFunctions["hotcue_" + i + "_enabled"] = "PioneerDDJSB3.hotCueLeds";
```

Comment:
It's pretty inefficient to update *all* hotcue LEDS when changing just a single hotcue.

Reply:
Reviewed this and this particular section is just setting up the mapping of the function to the control, it is not updating / changing the hotcues. This function runs once as part of the init routine.

hotCueLeds function itself only updates one hotcue when called.

Reply:
Doesn't look like it to me, it iterates over every single hotcue: https://github.com/mixxxdj/mixxx/pull/3821/files/838d17c172d0900ec200843af49479a9d667422c#diff-d5f946ca1b4972678ca7b6f9a6e80d25c1ad128ab7986944d89409e3ebd425c2R1222